File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Outtake.java
package org.firstinspires.ftc.teamcode;

import com.acmerobotics.dashboard.FtcDashboard;
import com.acmerobotics.dashboard.telemetry.TelemetryPacket;
import com.pedropathing.follower.Follower;
import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import com.qualcomm.robotcore.hardware.HardwareMap;
import org.firstinspires.ftc.teamcode.outtake.*;
import org.firstinspires.ftc.teamcode.utils.CancelableAction;
import org.firstinspires.ftc.teamcode.utils.Movement;
import org.jetbrains.annotations.NotNull;

/**
 * Class representing the Outtake mechanism of the robot.
 * Implements the CancelableAction interface to control various outtake components.
 */
public class Outtake implements CancelableAction {
    public Claw claw;
    public Pendulum pendulum;
    public Rotate rotate;
    /**
     * Constructor for the Outtake class.
     * Initializes all the outtake components using the provided hardware map.
     * @param hardwareMap the hardware map to use for initialization.
     */
    public Outtake(HardwareMap hardwareMap) {
        claw = new Claw(hardwareMap);
        pendulum = new Pendulum(hardwareMap);
        rotate = new Rotate(hardwareMap);
    }
    /**
     * Cancels all outtake actions.
     */
    @Override
    public void cancel() {
        claw.cancel();
        pendulum.cancel();
        rotate.cancel();
    }
    /**
     * Runs the outtake mechanism and updates the telemetry packet.
     * @param packet the telemetry packet to update.
     * @return true if all components are running, false otherwise.
     */
    @Override
    public boolean run(@NotNull TelemetryPacket packet) {
        return claw.run(packet) && pendulum.run(packet) && rotate.run(packet);
    }
    private OuttakePositions targetPosition;
    /**
     * Sets the target position for the outtake mechanism.
     * @param targetPosition the target position to set.
     */
    public void setTargetPosition(@NotNull OuttakePositions targetPosition) {
        this.targetPosition = targetPosition;
        pendulum.setTargetPosition(targetPosition.getPendulum());
        rotate.setTargetPosition(targetPosition.getRotate());
    }
    /**
     * Gets the target position for the outtake mechanism.
     * @return the target position.
     */
    public OuttakePositions getTargetPosition() {
        return targetPosition;
    }
    /**
     * Sets the claw position for the outtake mechanism.
     * @param isClosed the position to set.
     */
    public void setClaw(boolean isClosed) {
        claw.set(isClosed);
    }
    /**
     * Checks if the claw is closed.
     * @return true if the claw is closed, false otherwise.
     */
    public boolean isClosed() {
        return claw.isClosed();
    }
}
/**
 * TeleOp mode for testing the Outtake mechanism.
 * This class extends Movement and provides a simple teleop mode
 * to test the functionality of the Outtake mechanism.
 */
@TeleOp(name = "Outtake Test", group = "B")
class OuttakeTest extends Movement {
    private Outtake outtake;
    private FtcDashboard dash;
    @Override
    public void systemInit() {
        outtake = new Outtake(hardwareMap);
        dash = FtcDashboard.getInstance();
    }
    @Override
    public void systemLoop() {
        if (gamepad1.dpad_up) outtake.setTargetPosition(OuttakePositions.PICKUP);
        else if (gamepad1.dpad_down) outtake.setTargetPosition(OuttakePositions.TRANSFER);
        else if (gamepad1.dpad_left) outtake.setTargetPosition(OuttakePositions.BAR);
        else if (gamepad1.dpad_right) outtake.setTargetPosition(OuttakePositions.BASKET);

        if (gamepad1.right_bumper) outtake.setClaw(true);
        else if (gamepad1.left_bumper) outtake.setClaw(false);

        TelemetryPacket packet = new TelemetryPacket();
        if (!outtake.run(packet)) requestOpModeStop();
        dash.sendTelemetryPacket(packet);
    }
}
/**
 * TeleOp mode for manually controlling the Outtake mechanism.
 * This class extends LinearOpMode and provides a manual teleop mode
 * to control the Outtake mechanism using gamepad inputs.
 */
@TeleOp(name = "Manual Outtake Test", group = "C")
class OuttakeManual extends LinearOpMode {
    @Override
    public void runOpMode() {
        Outtake outtake = new Outtake(hardwareMap);
        FtcDashboard dashboard = FtcDashboard.getInstance();
        outtake.setTargetPosition(OuttakePositions.TRANSFER);

        waitForStart();
        double rotate = outtake.rotate.getTargetPosition();
        double pendulum = outtake.pendulum.getTargetPosition();

        while (opModeIsActive()) {
            if (gamepad1.right_bumper) outtake.setClaw(true);
            else if (gamepad1.left_bumper) outtake.setClaw(false);

            rotate += outtake.rotate.getMultiplier() * gamepad1.right_stick_y;
            pendulum += outtake.pendulum.getMultiplier() * gamepad1.left_stick_y;

            outtake.rotate.setTargetPosition(rotate);
            outtake.pendulum.setTargetPosition(pendulum);

            TelemetryPacket packet = new TelemetryPacket();
            if (!outtake.run(packet)) requestOpModeStop();
            dashboard.sendTelemetryPacket(packet);
        }
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/Outtake.kt
package org.firstinspires.ftc

import com.acmerobotics.dashboard.telemetry.TelemetryPacket
import com.qualcomm.robotcore.eventloop.opmode.Disabled
import com.qualcomm.robotcore.eventloop.opmode.TeleOp
import com.qualcomm.robotcore.hardware.HardwareMap
import org.firstinspires.ftc.outtake.Claw
import org.firstinspires.ftc.outtake.Pendulum
import org.firstinspires.ftc.outtake.Lift
import org.firstinspires.ftc.outtake.Rotate
import org.firstinspires.ftc.utils.CancelableAction
import org.firstinspires.ftc.utils.ManualPositionMechanism
import org.firstinspires.ftc.utils.Movement
import org.firstinspires.ftc.utils.PressAction

class Outtake(
    hardwareMap: HardwareMap,
    val claw: Claw = Claw(hardwareMap),
    val lift: Lift = Lift(hardwareMap),
    val pendulum: Pendulum = Pendulum(hardwareMap),
    val rotate: Rotate = Rotate(hardwareMap)
) : CancelableAction {
    /**
     * The list of outtake components.
     */
    private val components: MutableList<ManualPositionMechanism> = mutableListOf(claw, lift, pendulum, rotate)
    /**
     * Cancels all outtake actions.
     */
    override fun cancel() {
        components.forEach { it.cancel() }
    }
    /**
     * Runs the outtake mechanism and updates the telemetry packet.
     * @param p the telemetry packet to update.
     * @return true if all components are running, false otherwise.
     */
    override fun run(p: TelemetryPacket): Boolean = components.all { it.run(p) }
    /**
     * The target position for the outtake mechanism.
     */
    var targetPosition: OuttakePositions = OuttakePositions.TRANSFER
        set(value) {
            claw.targetPosition = value.rotate
            lift.targetPosition = value.lift
            pendulum.targetPosition = value.pendulum
            rotate.targetPosition = value.rotate
            field = value
        }
}
/**
 * Enum representing the different positions of the outtake mechanism.
 * Each position includes values for lift, pendulum and rotate components.
 */
enum class OuttakePositions (
    val lift: Double,
    val pendulum: Double,
    val rotate: Double
) {
    /**
     * Position for picking up specimens.
     * Includes values for the lift, pendulum and rotate components.
     */
    PICKUP(
        Lift.down,
        Pendulum.pickup,
        Rotate.pickup
    ),
    /**
     * Position for transferring samples to the outtake.
     * Includes values for the lift, pendulum and rotate components.
     */
    TRANSFER(
        Lift.transfer,
        Pendulum.transfer,
        Rotate.transfer
    ),
    /**
     * Position for rotating the outtake to the bar.
     * Includes values for the lift, pendulum and rotate components.
     */
    BAR(
        Lift.half,
        Pendulum.bar,
        Rotate.bar
    ),
    /**
     * Position for rotating the outtake to the basket.
     * Includes values for the lift, pendulum and rotate components.
     */
    BASKET(
        Lift.up,
        Pendulum.basket,
        Rotate.basket
    ),
    /**
     * Position for hanging the robot.
     * Includes values for the lift, pendulum and rotate components.
     */
    HANG(
        Lift.hang,
        Pendulum.transfer,
        Rotate.transfer
    )
}
/**
 * TeleOp mode for manually controlling the Outtake mechanism.
 * This class extends [Movement] and provides a manual teleop mode
 * to control the Outtake mechanism using gamepad inputs.
 */
@TeleOp(name = "Outtake Test", group = "B")
@Disabled
private class OuttakeTest : Movement() {
    /**
     * The Outtake mechanism to be tested.
     */
    override lateinit var system: Outtake
    /**
     * Initializes the Outtake mechanism.
     */
    override fun systemInit() {
        system = Outtake(hardwareMap)

        actions.add(PressAction(gamepad1::right_bumper) { system.claw.switch() })
    }
    /**
     * Updates the Outtake mechanism based on gamepad inputs.
     */
    override fun systemLoop() {
        system.pendulum.targetPosition += gamepad1.right_stick_y.toDouble() * system.pendulum.adjustMultiplier
        system.rotate.targetPosition += gamepad1.left_stick_y.toDouble() * system.rotate.adjustMultiplier
        system.lift.targetPosition += (gamepad1.right_trigger - gamepad1.left_trigger).toDouble() * system.lift.adjustMultiplier
    }
}
